csharp/src/Ice/ObjectAdapter.cs

```diff
@@ -876,36 +878,106 @@ public async Task SetPublishedEndpointsAsync(IEnumerable<Endpoint> newEndpoints)
             }
         }
 
-        internal ValueTask<OutgoingResponseFrame> DispatchAsync(IncomingRequestFrame request, Current current)
+        internal async ValueTask<OutgoingResponseFrame> DispatchAsync(
+            long streamId,
+            IncomingRequestFrame request,
+            Current current)
         {
-            Debug.Assert(current.Adapter == this);
-            IObject? servant = Find(current.Identity, current.Facet);
-            if (servant == null)
+            IDispatchObserver? dispatchObserver = null;
+            if (Communicator.Observer != null)
             {
-                throw new ObjectNotExistException(current.Identity, current.Facet, current.Operation);
+                dispatchObserver = Communicator.Observer.GetDispatchObserver(current, streamId, request.Size);
+                dispatchObserver?.Attach();
             }
 
-            return DispatchAsync(request, current, servant, 0);
+            try
+            {
+                Debug.Assert(current.Adapter == this);
+                IObject? servant = Find(current.Identity, current.Facet);
+                if (servant == null)
+                {
+                    throw new ObjectNotExistException(current.Identity, current.Facet, current.Operation);
+                }
+
+                OutgoingResponseFrame response = await DispatchAsync(servant, 0).ConfigureAwait(false);
```

Comment:
Why do we pass servant as a parameter?
Should we use a different name for the local function, like PerformDispatchAsync?

Reply:
Good question... Visual Studio Code printed an error when I removed the servant parameter so I thought it was required for local variables but not local parameters. I'll try again because it appears to work with a sample program.

Reply:
Ok so passing servant as a parameter is required because servant is not declared at the same scope as DispatchAsync. I've moved the definition before its use DispatchAsync as follow:
```
                ValueTask<OutgoingResponseFrame> DispatchAsync(int i)
                {
                    if (i < Interceptors.Count)
                    {
                        DispatchInterceptor interceptor = Interceptors[i++];
                        return interceptor(request, current, (request, current) => DispatchAsync(i));
                    }
                    else
                    {
                        return servant.DispatchAsync(request, current);
                    }
                }

                OutgoingResponseFrame response = await DispatchAsync(0).ConfigureAwait(false);
```